server/app/api/v1/charity/raffle/controller.js

```diff
@@ -0,0 +1,218 @@
+import { NotFoundError, NotPermittedError, ValidationError } from 'common-errors';
+import Charity from '../model';
+import Raffle from './model';
+import User from '../../user/model';
+
+import { VALID_OBJECT_ID } from '../../../../util/validations';
+
+
+/* *********************
+ * EXPORTED FUNCTIONS *
+ * *********************/
+
+/**
+ * Params middleware for any route with /raffle/:id
+ * sets req.raffle
+ */
+const params = (req, res, next, id) => {
+  Raffle.findById(id)
+    .select('-charity')
+    .populate('winner', 'email firstName lastName')
+    .exec()
+    .then((raffle) => {
+      if (!raffle) {
+        throw new NotFoundError('Raffle not found');
+      } else {
+        req.raffle = raffle;
+        next();
+      }
+    })
+    .catch(err => next(err));
+};
+
+/**
+ * Returns all raffles
+ * takes optional query isActive=true
+ */
+const all = (req, res, next) => {
+  if (req.query.hasOwnProperty('isActive') && req.query.isActive === 'true') {
+    return Raffle.find({ isActive: true })
+      .select('-charity')
+      .populate('winner', 'email firstName lastName')
+      .exec()
+      .then(raffles => res.json(raffles))
+      .catch(err => next(err));
+  }
+  return Charity
+    .populate(req.charity, {
+      path: 'raffles',
+      select: '-charity'
+    })
+    .then(({ raffles }) => {
+      return Raffle.populate(raffles, {
+        path: 'winner'
+      });
+    })
+    .then(raffles => res.json(raffles))
+    .catch(err => next(err));
+};
+
+/**
+ * Returns req.raffle
+ */
+const get = (req, res) => res.json(req.raffle);
+
+/**
+ * Creates new raffle
+ */
+const create = (req, res, next) => {
+  return new Raffle({
+    charity: req.charity._id,
+    prize: {
+      name: req.body.prizeName,
+      thumbnailURL: req.body.prizeThumbnailURL
+    }
+  }).save()
+  .then((savedRaffle) => {
+    req.charity.raffles.push(savedRaffle);
+    req.charity.save();
+    savedRaffle = savedRaffle.toObject();
+    return res.json(savedRaffle);
+  })
+  .catch(err => next(err));
+};
+
+/**
+ * Purchase raffle
+ * if existing user use query userId={userId}
+ * if new user user email, firstName, lastName in req.body
+ */
+
+const purchase = (req, res, next) => {
+  const hasUserId = req.query.hasOwnProperty('userId');
+  if (req.raffle.isActive) {
+    const findOrCreateUser = hasUserId
+      ? findByUserId
+      : createUser;
+
+    findOrCreateUser()
+      .then(addTickets)
+        .all()
+      .then(([raffle, user]) => {
+        const buyer = raffle.purchases.filter(
+          purchase => purchase.buyer === user._id
+        )[0];
+        return res.json({
+          count: buyer.count,
+          user: {
+            email: user.email,
+            firstName: user.firstName,
+            lastName: user.lastName
+          }
+        });
+      })
+      .catch(err => next(err));
+  } else {
+    next(new NotPermittedError(
+      'Tickets cannot be purchased after raffle has ended.'
```

Comment:
Need to verify if it's sending with a `message` property, but I'll change it to that if it's not. I forgot since the last time I tested it.